sycl/doc/extensions/proposed/sycl_ext_oneapi_kernel_compiler.asciidoc

```diff
@@ -0,0 +1,502 @@
+of `kernel_id` objects if the kernel bundle was created from a bundle of state
+`bundle_state::ext_oneapi_source`.
+
+Depending on the source language, the following kernel bundle functions may be
+used to query for a kernel when the kernel bundle was created from a bundle of
+state `bundle_state::ext_oneapi_source`.
+
+```
+template <typename KernelName>
+bool has_kernel() const noexcept;
+
+template <typename KernelName>
+bool has_kernel(const device& dev) const noexcept;
+
+template <typename KernelName>
+kernel get_kernel() const;
+```
+
+The specification for each source language tells how these functions behave and
+how the `KernelName` is interpreted.
+
+=== New kernel bundle member functions
+
+This extensions adds the following new `kernel_bundle` member functions:
+
+```
+namespace sycl {
+
+template <bundle_state State>
+class kernel_bundle {
+  // ...
+
+  // Available only if bundle_state is not bundle_state::ext_oneapi_source
+  bool ext_oneapi_has_kernel(const std::string &name);
+
+  // Available only if bundle_state is bundle_state::executable
+  kernel ext_oneapi_get_kernel(const std::string &name);
+};
+
+} // namespace sycl
+```
+
+|====
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+bool ext_oneapi_has_kernel(const std::string &name)
+----
+!====
+
+_Constraints:_ This function is not available when `State` is
+`bundle_state::ext_oneapi_source`.
+
+_Returns:_ The value `true` only if the kernel bundle was created from a bundle
+of state `bundle_state::ext_oneapi_source` and if it defines a kernel whose
+name is `name`.
+The extension specification for each source language tells how the `name`
+string is correlated to kernels defined in that source language.
+
+a|
+[frame=all,grid=none]
+!====
+a!
+[source]
+----
+kernel ext_oneapi_get_kernel(const std::string &name)
+----
+!====
+
+_Constraints:_ This function is available only when `State` is
+`bundle_state::executable`.
+
+_Returns:_ A `kernel` object representing the kernel in this bundle whose name
+is `name`.
+
+_Throws:_
+
+* An `exception` with the `errc::invalid` error code if
+  `ext_oneapi_has_kernel(name)` returns `false`.
+|====
+
+
+== Example
+
+The following example demonstrates how a SYCL application can define a kernel
+as a string and then compile and launch it.
+
+```
+#include <sycl/sycl.hpp>
+namespace syclex = sycl::ext::oneapi::experimental;
+
+int main() {
+  sycl::queue q;
+
+  // The source code for one or more kernels, defined in one of
+  // the supported source languages.
+  std::string source = R"""(
+    /* language specific kernel source code */
+  )""";
+
+  // Create a kernel bundle in "source" state.  The "some-language" is
+  // a stand-in for the enumerator telling which source language is used.
+  sycl::kernel_bundle<sycl::bundle_state::ext_oneapi_source> kb_src =
+    syclex::create_kernel_bundle_from_source(
+      q.get_context(),
+      syclex::source_language::/*some-language*/,
```

Comment:
Will the kernel compiler be able to compile ESIMD code?

Reply:
When the source language is "sycl", we will use the normal DPC++ compiler to compile the string.  Therefore, I think we can support all DPC++ features, including ESIMD.

Reply:
I think this comment is resolved, so I'm marking it that way.